### PR TITLE
avm2: Convert `has_class_in_chain` to use a `GcCell<Class>`

### PR DESCRIPTION
**Pull request by Aaron1011**
This lets us avoid looking up a `ClassObject` when performing parameter type coercion - we can instead use a `GcCell<Read>`. As a result, we no longer need all of the special logic for handling coercions during a script initializer.

### CODE REVIEW

**Comment by adrian17** [comment]
I'm not sure if there's anything specific here that'd require a new test; this refactor should be decently covered by tests already, right?

**Comment by Aaron1011** [comment]
Yeah, we have existing tests around the coercion logic and class lookup.

**Comment by Lord-McSweeney** [comment]
You stated that this progresses #2521- could you add a test for the issue it fixes in the SWF?

**Comment by Aaron1011** [comment]
I was mistaken - the progression is actually due to my PR implementing more methods in Transform, which lets the verifier pass when a class overrides them. Previously, this error was getting swallowed by the special-cased coercion logic.

**Review by adrian17 — APPROVED** [review]
Still not a fan of the diff, but let's just leave it for future us.

**Comment by adrian17** [comment]
note: on Discord I've found an edge case regarding Vectors. Avmplus does some fancier stuff there, but I think our solution for now could just copy `resolve_type`:
```rs
        if !type_name.params().is_empty() {
            let mut param_types = Vec::with_capacity(type_name.params().len());

            for param in type_name.params() {
                param_types.push(match self.resolve_type(param)? {
                    Some(o) => Value::Object(o.into()),
                    None => Value::Null,
                });
            }

            return Ok(Some(class.apply(self, &param_types[..])?));
        }
```
just with resolve_type replaced by domain class lookup (...and maybe no silly Vecs? :D )

File: core/src/avm2/property.rs [hunk 1]
```diff
@@ -64,56 +65,24 @@ impl<'gc> PropertyClass<'gc> {
...
-                let class = match outcome {
-                    ResolveOutcome::Class(class) => {
+                if name.is_any_name() {
+                    *self = PropertyClass::Any;
+                    (None, true)
+                } else {
+                    // Note - we look up the class in the domain by name, which allows us to look up private classes.
+                    // This also has the advantage of letting us coerce to a class while the `ClassObject`
+                    // is still being constructed (since the `Class` will already exist in the domain)
+                    let domain = unit.map_or(activation.avm2().stage_domain, |u| u.domain());
```
Comment (adrian17) [inline core/src/avm2/property.rs:75]: correct me if I'm wrong: I think `unit` can only be `None` if this comes from a builtin Rust-defined trait, so I think this should use a builtin domain instead?

**Review by adrian17 — APPROVED** [review]

File: core/src/avm2/property.rs (hunk 1 above)
Comment (Aaron1011) [reply core/src/avm2/property.rs]: You're right - I've changed it.

**Pull request merged**